keel-bakery/src/main/kotlin/com/netflix/spinnaker/keel/bakery/resource/ImageHandler.kt

```diff
@@ -0,0 +1,131 @@
+    "image",
+    "images"
+  ) to ImageSpec::class.java
+
+  override fun generateName(spec: ImageSpec): ResourceName =
+    ResourceName("bakery:image:${spec.artifactName}")
+
+  override fun desired(resource: Resource<ImageSpec>): Image =
+    with(resource.spec) {
+      val latestVersion = artifactRepository
+        .versions(DeliveryArtifact(artifactName, DEB))
+        .first()
+      Image(
+        baseAmiVersion = baseImageCache.getBaseImage(baseOs, baseLabel),
+        // TODO: there must be a canonical way of composing this
+        appVersion = latestVersion,
+        regions = regions
+      )
+    }
+
+  override fun current(resource: Resource<ImageSpec>): Image? =
+    runBlocking {
+      cloudDriver.images("aws", resource.spec.artifactName)
+        .await()
+        .firstOrNull()
+        ?.let {
+          // TODO: this is pretty crude to say the least
+          val tags = it.tagsByImageId.values.first()!!
+          Image(
+            tags.getValue("base_ami_version")!!,
+            tags.getValue("appversion")!!,
+            it.amis.keys
+          )
+        }
+    }
+
+  override fun upsert(
+    resource: Resource<ImageSpec>,
+    resourceDiff: ResourceDiff<Image>?
+  ): List<TaskRef> {
+    val taskRef = runBlocking {
+      val (_, application, version) = Regex("([\\w_]+)-(.+?)/.*")
+        .find(resourceDiff!!.source.appVersion)!!.groupValues
+      val artifact = igorService.getArtifact(application, version)
+      log.info("baking new image for {}", resource.spec.artifactName)
+      orcaService.orchestrate(
+        OrchestrationRequest(
+          name = "Bake image for ${resource.spec.artifactName}",
+          application = resource.spec.artifactName, // TODO: is this always the same? Does it matter?
+          description = "Bake image for ${resource.spec.artifactName}",
+          job = listOf(
+            Job(
+              "bake",
+              mapOf(
+                "amiSuffix" to "",
+                "baseOs" to resource.spec.baseOs,
+                "baseLabel" to resource.spec.baseLabel.name.toLowerCase(),
+                "cloudProviderType" to "aws",
+                "package" to resource.spec.artifactName,
+                "regions" to resource.spec.regions,
+                "storeType" to resource.spec.storeType.name.toLowerCase(),
+                "user" to "keel",
+                "vmType" to "hvm"
+              )
+            )
+          ),
+          trigger = OrchestrationTrigger(
+            correlationId = resource.metadata.name.toString(),
+            artifacts = listOf(artifact.await())
+          )
+        )
+      )
+        .await()
+    }
+    return listOf(TaskRef(taskRef.ref)) // TODO: wow, this is ugly
+  }
+
+  override fun delete(resource: Resource<ImageSpec>) {
+    TODO("not implemented")
+  }
+
+  override val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
+}
+
+/**
+ * The resolved representation of an [ImageSpec].
+ */
+data class Image(
```

Comment:
Represents the current / desired state when resolved against whatever versions of the package and base AMI exist